src/task.rs

```diff
@@ -480,81 +742,186 @@ enum State {
 
     NotifyStarlingErrored(Error, futures::sink::Send<mpsc::Sender<StarlingMessage>>),
     NotifyParentErrored(futures::sink::Send<mpsc::Sender<TaskMessage>>),
+
+    ShutdownChildren(Box<Future<Item = (), Error = Error>>),
 }
 
 enum NextState {
     EvaluateTopLevel(String),
+    HandleChildFinished(TaskId),
+    HandleChildErrored(TaskId, Error),
     NotifyStarlingFinished,
     NotifyStarlingErrored(Error),
     NotifyParentFinished,
     NotifyParentErrored(Error),
+    ShutdownChildren,
 }
 
 impl Future for Task {
     type Item = ();
     type Error = Error;
 
     fn poll(&mut self) -> Result<Async<Self::Item>> {
+        // Principals for error handling, recovery, and propagation when driving
+        // tasks to completion:
+        //
+        // * Whenever possible, catch errors and then inform
+        //     1. the Starling system supervisor thread,
+        //     2. the parent task (if any),
+        //     3. and this task's children
+        //   in that order to propagate them.
+        //
+        // * Tasks should never outlive the Starling system supervisor thread,
+        //   so don't attempt to catch errors sending messages to it. Just panic
+        //   if that fails, since something very wrong has happened.
+        //
+        // * We can't rely on this task's children or parent stricly observing
+        //   the tree lifetime discipline that we expose to JavaScript. There
+        //   can be races where both shut down at the same time, perhaps
+        //   involving the main task returning and the Starling system
+        //   supervisor thread sending shutdown notices to all remaining
+        //   tasks. Therefore, we allow all inter-task messages to fail
+        //   silently, locally ignoring errors, and relying on the Starling
+        //   system supervisor thread to perform ultimate clean up.
```

Comment:
Do we have any guarantees that this really only happens in corner cases around race conditions? It'd be nice not to be overly generous in dropping things on the floor here.

Reply:
Sending a message can only fail if the receiver has been dropped, so at that point we know that the other task has shutdown, and I'm not really sure how to best react to that. I can look into tightening inter-task communication up.